[util] Add script for computing contributor statistics

AlexJones0 opened the pull request:
Supersedes #27173, extending it to support pagination. CC @timothytrippel (thank you for the original script!)

I've also added a few more features to the script, mainly:
  - The ability to filter all historical committers from the results with `--filter`, so that the results only include those who have never been featured in the `COMMITTERS` file.
  - Optional progress logging, as the pagination and size of the repo means that the search takes a while (~5 minutes).
  - Options to limit the number of pages to fetch, so the script can be more easily iterated on and tested.

If you want to run this script to test, I recommend running it as `./util/py/scripts/committer_stats.py -l -f` (removing `-f` if you want to see the details for *all* historical contributors).

**Note**: the last two commits contain changes to the original logic, to ensure that (a) reviewers are only counted once per PR, and (b) authors of PRs are not counted as reviewers for their own PRs. This will mean that the review counts reported by the new script will be lower than before (but will be accurate) - thus any interpretation of these numbers compared to older figures should probably take this into account.

andreaskurth commented:
Thanks for your work on this, @AlexJones0.

I just tried running the script to check it's results.

Without modifications, I get `requests.exceptions.ConnectTimeout: HTTPSConnectionPool(host='api.github.com', port=443): Max retries exceeded with url: /graphql (Caused by ConnectTimeoutError(<urllib3.connection.HTTPSConnection object at 0x7fe5c4e67d90>, 'Connection to api.github.com timed out. (connect timeout=None)'))` at `Searched 13600 commits for author information...`.

I'm able to bypass this problem with
```diff
diff --git a/util/py/scripts/committer_stats.py b/util/py/scripts/committer_stats.py
index 8b411433b7..3123613949 100755
--- a/util/py/scripts/committer_stats.py
+++ b/util/py/scripts/committer_stats.py
@@ -72,6 +72,7 @@ def run_github_graphql_query(
 
     while True:
         response = requests.post(GH_GRAPHQL_API_ENDPOINT, json=payload, headers=headers)
+        time.sleep(2)
         response.raise_for_status()
 
         # Check status code
```
but I think the script should prevent that without a general slowdown, if possible.

More importantly, I think the results aren't correct yet. Take `martin-velay` as example, who [currently has 99 PRs merged to `master` ](https://github.com/lowRISC/opentitan/pulls?q=is%3Amerged+is%3Apr+author%3Amartin-velay+base%3Amaster) and who [currently has reviewed 58 PRs to `master` that he hasn't authored himself](https://github.com/lowRISC/opentitan/pulls?q=is%3Apr+reviewed-by%3Amartin-velay+base%3Amaster+-author%3Amartin-velay). The current version of the script gives very different results for him:
```
Pull Request Counts per Contributor:
[...]
- martin-velay: 8
[...]

Review Counts per Reviewer:
[...]
- martin-velay: 1
```

Finally, would it be possible to format the output similar to this list:
| Account name | Commits | Merged PRs | PRs reviewed |
| --- | --- | --- | --- |
| martin-velay | 136 | 99 | 58 |

and have the list sorted by commits, then merged PRs, then PRs reviewed?

Thanks in advance! :-)

AlexJones0 commented:
> Without modifications, I get requests.exceptions.ConnectTimeout: HTTPSConnectionPool(host='api.github.com', port=443): Max retries exceeded with url: /graphql (Caused by ConnectTimeoutError(<urllib3.connection.HTTPSConnection object at 0x7fe5c4e67d90>, 'Connection to api.github.com timed out. (connect timeout=None)')) at Searched 13600 commits for author information....

Interesting! I'm not seeing this locally. The error message makes it sound like your device is not able to reliably connect to the Github API graphQL endpoint, so maybe their services were having some issues? I can look at adding some timing or another workaround here if we think it will make it more stable though.

> More importantly, I think the results aren't correct yet.

Indeed - I think my most recent [force push](https://github.com/lowRISC/opentitan/compare/b347e74089dac54c7401dceaa0929012cc280abd..6a71a689448ad389e21ea26daf4642950caf3c57) broke it since I was seeing correct numbers before, but I realised that I had missed some logic of the original script which was only querying closed PRs. I think that maybe this original PR was wrong, and `CLOSED` refers instead to PRs that were closed without merging?

More generally, looking at your output, it seems the desired goal is:
  - Authored PRs should only count if merged.
  - Reviews should count even if the PR is not merged.
If so, I can try and change the logic to match this.

> Finally, would it be possible to format the output similar to this list:

Sure - I'll make that change now. Do you mean to use merged PRs and reviews as a tie-breaker when sorting, or to display three repeated lists, each sorted by a different column?

AlexJones0 commented:
@andreaskurth  Would you be able to test it again for me? I've added some countermeasures that should hopefully catch issues with poorer connections and conditionally add the timing delay that you suggested. I've also fixed the `[CLOSED]` logic issue I described and made sure the three fields match your descriptions. I wasn't sure whether you wanted one or three tables, so I've just put three for now (though the first table is sorted by commits, then merged PRs, then reviews anyway). We can just drop these last 2 tables if they weren't what you wanted.

When I run it locally I get:
```
Contributors sorted by authored commits:
+-------------------------------+---------+------------+--------------+
|         Account name          | Commits | Merged PRs | PRs Reviewed |
+-------------------------------+---------+------------+--------------+
               [...]
|         martin-velay          |   136   |     99     |      58      |
               [...]
|          AlexJones0           |   86    |     43     |      37      |
               [...]
```

which seems to match what you suggest, as well as my stats for [merged PRs](https://github.com/lowRISC/opentitan/pulls?q=is%3Amerged+is%3Apr+author%3Aalexjones0+base%3Amaster) and [reviews](https://github.com/lowRISC/opentitan/pulls?q=is%3Apr+reviewed-by%3Aalexjones0+base%3Amaster+-author%3Aalexjones0).

andreaskurth commented:
Thanks @AlexJones0! The API querying works reliably for me now, and the results seem to match those of the respective web queries. :+1:

Re output format: I think the sorted table that currently gets printed first (i.e., rows sorted by commits, then (to break ties) merged PRs, then reviewed PRs) suffices. The other tables can be removed from the output. If different sorting is needed, I think it's simple enough to tweak the Python code, and if really needed a sorting argument could be added later.

Would it be possible to move "Unknown Author" to the bottom of the table, though?

And would it be possible to optionally (with an argument, e.g., `--format=html`) output the table as HTML?

AlexJones0 commented:
@andreaskurth That's great to hear.

I have limited the output to just the first table and moved "Unknown Author" to the bottom of the table. I have added a `-F`/`--format` option which takes a default value of `table` as in the previous force push, but also accepts `html_output` to print the HTML source for rendering a basic table, and `html_display` which actually writes that HTML to a file and opens it in a web browser. I added both of these since I wasn't sure if you were asking for just the HTML source, or you wanted it to render the HTML in a browser. If you prefer just one of these let me know and we can just use that option.

There could maybe be some more work done (the HTML styling is basic, the script is more complex than when initially designed and so could be refactored a bit), but I am out-of-office this week so further changes may have to wait.
